tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_shape_calc_placement.cc

```diff
@@ -0,0 +1,539 @@
+      {"mhlo.dynamic_conv", {/*paddings*/ 2}},
+      {"mhlo.if", {/*pred*/ 0}},
+      {"mhlo.dynamic_rng_uniform", {/*start*/ 0, /*limit*/ 1, /*shape*/ 2}}};
+
+  // add output OP into marked set if it is a I64 scalar and placment is CPU.
+  void markI64ReturnedCPUScalarOps(FuncOp func,
+                                   DenseMap<Operation*, bool>& marked_ops);
+  // Update marked set.
+  // If a OP is in marked set, add all of its operands to marked set.
+  // Add some operands of dynamic shape OPs into marked set according to lookup
+  // table.
+  void markShapeCalculationOps(FuncOp func,
+                               DenseMap<Operation*, bool>& marked_ops);
+
+  // Get placement vector of func's output.
+  SmallVector<llvm::StringRef, 4> getOutputPlacements(FuncOp main_func);
+
+  // Get Op's placement according to its attr
+  PlacementType getOpPlacement(Operation* op);
+
+  // Get tensor's placement
+  PlacementType getTensorPlacement(Operation* dst, size_t operand_idx);
+
+  // Get input argument's placment.
+  PlacementType getArgumentPlacement(Value arg);
+
+  // Enforce output's placement.
+  void enforceOutputPlacement(
+      Operation* dst, FuncOp main_func,
+      SmallVector<std::pair<Operation*, size_t>, 4>& d2h_worklist,
+      SmallVector<std::pair<Operation*, size_t>, 4>& h2d_worklist);
+
+  // insert H2D or D2H Op.
+  void insertMemcpy(Operation* dst, size_t operand_index, bool is_h2d);
+};
+
+// Place the subgraph that is the producer of any shape operands
+// on Host
+// TODO(disc): handle when TupleOp exists in shape_calc_ops
+void PlaceShapeCalcOnHost::placeShapeCalcSubgraphOnHost() {
+  ModuleOp module = getOperation();
+  Builder builder(&getContext());
+  llvm::DenseMap<Operation*, bool> shape_calc_ops;
+
+  for (FuncOp func : module.getOps<FuncOp>()) {
+    // Put the i64 Scalar output on Host(into shape_calc_ops).
+    // TODO(disc): revisit this if we have outputs on CPU for TF in the future.
+    if (func.getName() == "main") {
+      markI64ReturnedCPUScalarOps(func, shape_calc_ops);
+    }
+    // skip empty blocks
+    if (func.getBlocks().size() == 0) continue;
+    // no target ops
+    if (llvm::none_of(func.getBlocks().front(), [](Operation& op) {
+          auto dialect_name = op.getDialect()->getNamespace();
+          return isTargetDialect(dialect_name);
+        })) {
+      continue;
+    }
+    markShapeCalculationOps(func, shape_calc_ops);
+  }
+
+  for (auto op : shape_calc_ops) {
+    if (op.second) {
+      // We suppose that mhlo op only has single output, either having tensor
+      // type or tuple type.
+      if (auto tp = op.first->getResult(0).getType().dyn_cast<TupleType>()) {
+        // If an op is placed on host, then we suppose all its outputs are
+        // placed on host.
+        SmallVector<Attribute, 4> attrs(tp.size(),
+                                        builder.getStringAttr(kTypeHost));
+        op.first->setAttr(kPlaceTyAttr, ArrayAttr::get(tp.getContext(), attrs));
```

Comment:
Can we just consider the Tuple as a single unique and mark the placement with a single attribute instead of an array?
So that we have one placement per SSA value?
(this is already what you're doing anyway)

Reply:
No, we cannot assume that. The elements in tuple could be placed on difference devices. For example, tuple is the producer of while op, we cannot limit its elements on same device.

Reply:
Mmmm, OK we're still having some XLA legacy use of tuple here. I think we'd like to get rid of tuples entirely but we're not there yet. So LG for now.